src/libsystemd/sd-hwdb/sd-hwdb.c

```diff
@@ -319,23 +319,23 @@ _public_ int sd_hwdb_new(sd_hwdb **ret) {
 
         /* find hwdb.bin in hwdb_bin_paths */
         NULSTR_FOREACH(hwdb_bin_path, hwdb_bin_paths) {
+                log_debug("Trying to open \"%s\"...", hwdb_bin_path);
                 hwdb->f = fopen(hwdb_bin_path, "re");
                 if (hwdb->f)
                         break;
-                else if (errno == ENOENT)
-                        continue;
-                else
+                if (errno != ENOENT)
                         return log_debug_errno(errno, "Failed to open %s: %m", hwdb_bin_path);
         }
 
-        if (!hwdb->f) {
-                log_debug("hwdb.bin does not exist, please run 'systemd-hwdb update'");
-                return -ENOENT;
-        }
+        if (!hwdb->f)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOENT),
+                                       "hwdb.bin does not exist, please run 'systemd-hwdb update'");
 
-        if (fstat(fileno(hwdb->f), &hwdb->st) < 0 ||
-            (size_t) hwdb->st.st_size < offsetof(struct trie_header_f, strings_len) + 8)
-                return log_debug_errno(errno, "Failed to read %s: %m", hwdb_bin_path);
+        if (fstat(fileno(hwdb->f), &hwdb->st) < 0)
+                return log_debug_errno(errno, "Failed to stat %s: %m", hwdb_bin_path);
+        if ((size_t) hwdb->st.st_size < offsetof(struct trie_header_f, strings_len) + 8)
```

Comment:
this size check is weird btw, and I think should be fixed too

on 32bit, size_t is 32bit, but .st_size is off_t hence 64bit. Now you can this larger type to the smaller one, before comparing. This should be done the other way, i.e. the result of offsetof() should be cast to off_t before comparing. 

(not an issue introduced by your PR, but if you touch this, please fix)